polytracker/mapping.py

```diff
@@ -3,156 +3,191 @@
-            for interval in sorted(unused_bytes):
-                yield TaintedRegion(
-                    source=source,
-                    offset=interval.begin,
-                    length=interval.end - interval.begin,
-                )
+    def __init__(self, f: TDFile):
+        self.tdfile: TDFile = f
+
+    def mapping(self) -> Dict[OffsetType, Set[OffsetType]]:
+        raise NotImplementedError()
+
+    def marker_to_ranges(self, m: bytearray) -> List[CavityType]:
+        ranges = []
+        start = None
+        for i, v in enumerate(m):
+            if v == 0:
+                if start is None:
+                    start = i
+            if v == 1:
+                if start is not None:
+                    ranges.append((start, i))
+                    start = None
+        if start is not None:
+            ranges.append((start, len(m) - 1))
+        return ranges
+
+    def file_cavities(self) -> Dict[Path, List[CavityType]]:
+        seen: Set[int] = set()
+
+        def source_labels_not_affecting_cf(label: int) -> Generator[int, None, None]:
+            stack = [label]
+            while len(stack) > 0:
+                lbl = stack.pop()
+
+                if lbl in seen:
+                    continue
+
+                seen.add(lbl)
+
+                n = self.tdfile.decode_node(lbl)
+
+                if n.affects_control_flow:
+                    continue
+
+                if isinstance(n, TDSourceNode):
+                    yield lbl
+
+                elif isinstance(n, TDUnionNode):
+                    nl = self.tdfile.decode_node(n.left)
+                    if not nl.affects_control_flow:
+                        if isinstance(nl, TDSourceNode):
+                            yield n.left
+                        else:
+                            stack.append(n.left)
+
+                    nr = self.tdfile.decode_node(n.right)
+                    if not nr.affects_control_flow:
+                        if isinstance(nr, TDSourceNode):
+                            yield n.right
+                        else:
+                            stack.append(n.right)
+
+                elif isinstance(n, TDRangeNode):
+                    for rl in range(n.first, n.last + 1):
+                        # NOTE: One could skip decoding here, but then we could end up with really long ranges
+                        # being added the labels that really does nothing except cause overhead...
+                        rn = self.tdfile.decode_node(rl)
+                        if rn.affects_control_flow:
+                            continue
+                        if isinstance(rn, TDSourceNode):
+                            yield rl
+                        else:
+                            stack.append(rl)
+
+        result: Dict[Path, List[CavityType]] = defaultdict(list)
+
+        for p, h in self.tdfile.fd_headers:
+            begin = h.prealloc_label_begin
+            end = h.prealloc_label_end
```

Comment:
We previously had the discussion about not always preallocating ranges. Would it make sense to add an assert here of some sort? E.g. assert that `end != 0 and begin !=0`. I guess it might not work because output files are in this enumeration as well and would typically have initial size zero.

Reply:
There's a test for length on line 96 :eyes: Is it not enough?

Reply:
My point is that if we change the implementation to not require preallocated ranges this would stop working. Maybe that's what we have tests for...